[Pytorch] NVIDIA-DL-Framework-Inspect support – part 4 – documentation (#1611)
NVIDIA/TransformerEngine: docs/debug/1_getting_started.rst

..
    Copyright (c) 2022-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.

    See LICENSE for license information.

Getting started
==============

> .. warning::

Reviewer: I think a note is more appropriate here instead of a warning.

   Precision debug tools with `Nvidia-DL-Framework-Inspect <https://github.com/NVIDIA/nvidia-dlfw-inspect>`_ for Transformer Engine are currently supported only for PyTorch.

Transformer Engine provides a set of precision debug tools which allow you to easily:

- log the statistics for each of the tensors in every matrix multiply (GEMM) operation,
- run selected GEMMs in higher precision,
- run current scaling - with one scaling factor per tensor - for particular GEMMs,
- test new precisions and integrate them with FP8 training,

> - ... and many more, with only a few changes to the code.

Reviewer: ```suggestion
- ... and many more.
```
You already say "easily"

There are 4 things one needs to do to use Transformer Engine debug features:

> 1. Create a **config.yaml** file to configure the desired features.

Reviewer: Is the filename configurable somehow or does it have to be config.yaml?

2. Import, and initialize the `Nvidia-DL-Framework-Inspect <https://github.com/NVIDIA/nvidia-dlfw-inspect>`_ tool, which is installed as the dependency of the Transformer Engine.

> 3. One can pass ``name="..."`` to the init of every TE layer to easier identify layer names. If this is not provided, names will be inferred automatically.

Reviewer: ```suggestion
3. One can pass ``name="..."`` when creating TE layers to easier identify layer names. If this is not provided, names will be inferred automatically.
```

4. Invoke ``debug_api.step()`` at the end of one forward-backward pass.

To start debugging, one needs to create a ``config.yaml`` file. This file lists the features to be used in particular layers. There are 2 kinds of features:

- provided by the Transformer Engine - for example, DisableFP8GEMM or LogTensorStats - they are listed in the :doc:`debug features API <3_api_features>` section
- defined by the user. For details on how to create a custom feature - please read the :doc:`calls to Nvidia-DL-Framework-Inspect <3_api_te_calls>` section.

> .. figure:: ./img/introduction.svg

Reviewer: There are some issues with this Figure:
 - config.yaml says that seciton 2 applies to Linear1, but the rest of the image suggests it applies to Linear2
 - "Providedby" on the right side of the image should have the space between Provided and by

2. Logging statistics for other GEMM operations, such as gradient statistics for data gradient GEMM within the LayerNormLinear sub-layer of the TransformerLayer.

Config file
----------

We need to prepare the **config.yaml** file, as below

.. code-block:: yaml

    # config.yaml

    fc1_fprop_to_fp8:
      enabled: True
      layers:
        layer_types: [fc1, fc2] # contains fc1 or fc2 in name
      transformer_engine:
        DisableFP8GEMM:
          enabled: True
          gemms: [fprop]

    log_tensor_stats:
      enabled: True
      layers:
        layer_types: [layernorm_linear] # contains layernorm_linear in name
      transformer_engine:
        LogTensorStats:
          enabled: True
          stats: [max, min, mean, std, l1_norm]
          tensors: [activation]
          freq: 1
          start_step: 2
          end_step: 5

Further explanation on how to create config files is in the :doc:`next part of the documentation <2_config_file_structure>`.

Adjusting Python file
--------------------

.. code-block:: python

    # (...)

    import nvdlfw_inspect.api as debug_api
    debug_api.initialize(

>         config_file="./config.yaml",

Reviewer: This suggests that the name is configurable. Could we then, rather than using bold config.yaml everywhere, say something like "configuration YAML file" at least once to not confuse people? And then in this example we could say that we are creating the configuration file, which we named config.yaml.
